getting_started/Data_Input_Pipeline.md

```diff
@@ -15,12 +15,61 @@
+hf_data_files: 'gs://<bucket>/<folder>/*-train-*.parquet'  # match the train files
+# for HF pipeline, tokenizer_path can be a path in HuggingFace Hub, 
+# or a local path containing tokenizer in a format supported by transformers.AutoTokenizer
+tokenizer_path: 'google-t5/t5-large'  # for using https://huggingface.co/google-t5/t5-large
+```
+#### Limitations & Recommendations
+1. Streaming data directly from HuggingFace Hub may be impacted by the traffic of the server. During peak hours you may encounter "504 Server Error: Gateway Time-out". It's recommended to download the HuggingFace dataset to a GCS buckt or disk for the most stable experience.
+2. Streaming data directly from HuggingFace Hub works in multihost settings with a samll number of hosts. We have encountered "read time out" error with host number > 16.
+3. Multihost data input is achieved through HuggingFace's `datasets.distributed.split_dataset_by_node` API, which is more performant when (data shards/files) % (number of host reading data) == 0. To give MaxText users a performant experience, we follow this rule and will use data shards up to the number that's divisible by number of host reading data. For instance if your dataset has 100 shards, and you use 8 hosts, at most 96/100 shards of data will be used since 96%8==0. You can reshard or combine shards if you want to utilize the whole dataset.
+
+### Grain pipeline - for determinism
+
+#### Why do we need determinism for data input pipeline?
+Determinism in a data input pipeline means that the same input data always results in the same sequence of batches at each step. This is typically achieved by setting a fixed shuffle seed during pipeline initialization. In an ideal scenario, where training runs uninterrupted, this determinism is straightforward (deterministic without preemption). However, real-world distributed training environments often face preemptions due to maintenance, hardware failures, or resource constraints. 
+When a preempted training run resumes, the data input pipeline is re-initialized. If the same shuffle seed is used, the pipeline restarts from the beginning, potentially re-training the model on initial data. Conversely, a new seed produces a different batch sequence, making it difficult to track which data has been seen and how often each example is used for training. This lack of control can impact model performance and reproducibility.
+
+#### Cases where determinism is crucial
+* **Model sensitive to repetition.** When models are sensitive to the frequency with which they encounter specific examples, precise control over the order and repetition of data during training is essential.
+* **Convergence comparison.** In sensitive convergence experiments like testing quantization techniques, maintaining identical data batches between runs (e.g., quantized vs. unquantized) is essential for comparison. Determinism ensures consistency even the runs are long and undergo saving/resuming at different step.
```

Comment:
nit: consistency even **when** the runs are long

Reply:
Thanks for catching, fixed now.